src/components/Timer.react.js

```diff
@@ -0,0 +1,317 @@
+import PropTypes from 'prop-types';
+import React, {Component} from 'react'; // eslint-disable-line no-unused-vars
+import prettyMilliseconds from 'pretty-ms';
+
+function getFormat(format_type) {
+    const formats = {
+        default: {},
+        verbose: {verbose: true},
+        colons: {colonNotation: true},
+        compact: {compact: true},
+    };
+    return formats[format_type];
+}
+
+/**
+ * The Timer component has all the functionality of the Interval component plus
+ * the following additional features:
+ *
+ * Operate in either `countdown` or `stopwatch` (count up) modes.
+ * Display custom messages, or start/stop jobs at specified times.
+ * Convert milliseconds into human readable times.
+ * Update messages clientside to help improve app performance.
+ * Specify the elapsed times to fire a callback rather than every interval
+ **/
+
+export default class Timer extends Component {
+    constructor(props) {
+        super(props);
+        this.intervalId = null;
+        this.renderMessage = null;
+        this.reportInterval = this.reportInterval.bind(this);
+        this.handleTimer = this.handleTimer.bind(this);
+        this.handleMessages = this.handleMessages.bind(this);
+    }
+
+    handleTimer(props) {
+        const {
+            n_intervals,
+            max_intervals,
+            disabled,
+            rerun,
+            interval,
+            time,
+            mode,
+            duration,
+        } = this.props;
+
+        // Check if timer should stop or shouldn't even start
+        if (
+            max_intervals === 0 ||
+            disabled ||
+            (n_intervals >= max_intervals && max_intervals !== -1) ||
+            duration === 0 ||
+            duration < 0 ||
+            (mode === 'countdown' && duration === null) || // operates as stopwatch only if duration is null
+            (mode === 'countdown' && time === 0 && !rerun) ||
+            (mode === 'stopwatch' && time === duration && !rerun) ||
+            time < 0
+        ) {
+            // stop existing timer
+            if (this.intervalId) {
+                this.clearTimer();
+            }
+            // and don't start a timer
+            return;
+        }
+
+        // keep the existing timer running
+        if (this.intervalId) {
+            if (
+                (mode === 'countdown' && time === 0 && rerun) ||
+                (mode === 'stopwatch' && time === duration && rerun)
+            ) {
+                this.initTimer(props);
+            }
+            return;
+        }
+
+        // it hasn't started yet (& it should start)
+        this.intervalId = window.setInterval(this.reportInterval, interval);
+    } // end handle timer
+
+    handleMessages(props, new_time) {
+        const {messages, timer_format} = this.props;
+
+        const messagesObj = Object.assign({}, messages);
+        if (new_time in messagesObj) {
+            this.renderMessage = `${messagesObj[new_time]}`;
```

Comment:
```suggestion
            this.renderMessage = messagesObj[new_time];
```
Same simplification as below, I missed this one the first time around. Backtick template literals are great to avoid concatenations but they're just passthroughs here.